README.md
authbind
========
Allows non-root programs to bind() to low ports
This package allows a package to be started as non-root but
still bind to low ports, without any changes to the application.

http://en.wikipedia.org/wiki/Authbind


This github repo is a clone of the official repo: 
  http://www.chiark.greenend.org.uk/ucgi/~ian/git/authbind.git/

This version has an added SPEC file for building RPM.


debian/changelog
authbind (2.1.2~~iwj) unstable; urgency=low

  * Fix one-letter typo in previous changelog entry.

 --

authbind (2.1.1) unstable; urgency=low

  * Manpage has an example of which files will be checked and read
    Closes: #676540.
  * Do not include library version number in usage output from helper; the
    library and package version numbers are not the same, so this is
    misleading, and we do not want to introduce new machinery just to
    plumb the package version number through.  Closes: #676440.
  * Another minor wording fix in the manpage.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Sun, 10 Jun 2012 23:17:14 +0100

authbind (2.1.0) unstable; urgency=low

  * Permissions file lines can contain IPv6 mask/length.
  * Permissions file lines need only contain one port number.
  * Manpage improvements.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Sun, 03 Jun 2012 12:33:53 +0100

authbind (2.0.1) unstable; urgency=low

  Bugfix:
  * 0-length ipv4 address masks (ie, 0.0.0.0/0) in byuid files wouldn't
    work because they would attempt a 32-bit shift.  This was reported
    against Ubuntu in Launchpad but the but is here too.  LP: #514278.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Sun, 03 Jun 2012 01:50:13 +0100

authbind (2.0.0) unstable; urgency=medium

  Important new functionality:
  * Support IPv6.  (Closes: #596921.)
  * Support ports 512-1023 if the user really wants.
    Explain the problem more clearly in the manpage.  (Closes: 654706.)

  Bugfixes:
  * Correct manpage description of helper protocol to have
    actually-implemented byte order convention.  (Closes: #651694.)
  * Improve wording in BUGS section of manpage about forking.

  Packaging fixes:
  * Remove shlibs files.  These are not needed because nothing should
    actually link to libauthbind.
  * Update Standards-Version.
  * Upstream repo is now in git.
  * Change my email address.
  * Set Priority to optional as in ftpmaster override file.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Sat, 02 Jun 2012 01:56:37 +0100

authbind (1.2.0) unstable; urgency=low

  * Block signals during our child's execution.  Closes: #186439.
  * manpage: fix subtle error in description of interaction between
    set-id and LD_PRELOAD.  Closes: #36890.
  * Say in manpage that we may not work properly with multithreaded
    programs.  See also Bug#186439.
  * Make sure we don't try to encode errno's >127 in a status code.
  * New make variables for improved portability and customisation of
    locations: LIBS, libexec_dir, HELPER, BINTARGETS.
  * make clean removes *.core too.
  * #include <signal.h> in libauthbind.c.
  * Set SHELL=/bin/bash in debian/rules.  Closes: #374586.
  * Revert Amaya's accidental change to CC= line in debian/rules.
  * Debian-native again.  Closes: #393556.
  * Do not strip if DEB_BUILD_OPTIONS=nostrip (Closes: #436489):
    - in Makefile, do not set INSTALL_* or STRIP if already set
    - in rules, export STRIP=: if nostrip
  * Change my email address in Maintainer: to davenant.
    
 -- Ian Jackson <ian@davenant.greenend.org.uk>  Sun, 23 Sep 2007 20:23:50 +0100

authbind (1.1.5.2-0.2) unstable; urgency=low

  * Non-maintainer upload.
  * Get rid of the /usr/doc link (Closes: #359363).
  * Revamped packaging a bit:
    - Get rid of Build-Depends on libc6-dev (build-essential)
    - spelling-error-in-copyright Debian/GNU Linux Debian GNU/Linux
    - Not debian-native anymore
    - Updated ancient-standards-version to 3.7.2, no changes needed.

 -- Amaya Rodrigo Sastre <amaya@debian.org>  Wed, 12 Jul 2006 12:28:12 +0200

authbind (1.1.5.2-0.1) unstable; urgency=low

  * Non-maintainer Upload
  * Moved man pages to /usr/share/man and docs to
    /usr/share/doc/authbind. Added a prerm and a postinst script to set
    the /usr/doc/authbind link. (closes: #91112, #91376, #91387)
  * Fixed the path to the common licenses in debian/copyright
  * Added a section and priority field to the changelog
  * Added support for the `nostrip' and `debug' options in
    DEB_BUILD_OPTIONS.
  * Boosted Standards-Version to 3.5.2

 -- Bas Zoetekouw <bas@debian.org>  Sun, 13 May 2001 18:00:52 +0200

authbind (1.1.5.2) frozen unstable; urgency=low

  * Non-maintainer upload.
  * Makefile: Add -D_GNU_SOURCE, closes: 89944

 -- Stephen Frost <sfrost@debian.org>  Fri, 16 Mar 2001 19:42:05 -0500

authbind (1.1.5.1) frozen unstable; urgency=low

  * Non-maintainer upload.
  * debian/rules (clean): use tab not spaces, so `debian/rules clean'
    actually works. [#27360]
  * debian/rules (binary-arch): don't compress the copyright file.
  * Makefile: install the shared library non-executable.
 
 -- James Troup <james@nocrew.org>  Sun,  3 Jan 1999 21:23:50 +0000

authbind (1.1.5) unstable; urgency=low

  * Fix coredump bug !

 -- Ian Jackson <ian@davenant.greenend.org.uk>  Mon, 31 Aug 1998 15:32:44 +0100

authbind (1.1.4) unstable; urgency=low

  * Allow bind() to port 0 !

 -- Ian Jackson <ian@davenant.greenend.org.uk>  Mon, 31 Aug 1998 15:23:30 +0100

authbind (1.1.3) unstable; urgency=low

  * Make IPPORT_RESERVED check work (oops!).

 -- Ian Jackson <ian@davenant.greenend.org.uk>  Sun, 30 Aug 1998 20:52:10 +0100

authbind (1.1.2) unstable; urgency=low

  * Make --deep option actually work (oops).

 -- Ian Jackson <ian@davenant.greenend.org.uk>  Sun, 30 Aug 1998 20:09:39 +0100

authbind (1.1.1) unstable; urgency=low

  * Fix Description.

 -- Ian Jackson <ian@davenant.greenend.org.uk>  Sun, 30 Aug 1998 16:34:28 +0100

authbind (1.1) experimental; urgency=low

  * Allow control of when the LD_PRELOAD is passed on.

 -- Ian Jackson <ian@davenant.greenend.org.uk>  Sun, 30 Aug 1998 16:18:24 +0100

authbind (1.0) experimental; urgency=low

  * Following testing, we can call this 1.0.

 -- Ian Jackson <ian@davenant.greenend.org.uk>  Sun, 30 Aug 1998 03:15:21 +0100

authbind (0.1-1) experimental; urgency=low

  * Initial release.

 -- Ian Jackson <ian@davenant.greenend.org.uk>  Sat, 29 Aug 1998 20:10:37 +0100


helper.c
/*
 *  helper.c - setuid helper program for authbind
 *
 *  authbind is Copyright (C) 1998 Ian Jackson
 * 
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 * 
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software Foundation,
 *  Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. 
 * 
 */

#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <assert.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#ifndef CONFIGDIR
# define CONFIGDIR "/etc/authbind"
#endif

static void exiterrno(int e) {
  exit(e>0 && e<128 ? e : ENOSYS);
}

static void perrorfail(const char *m) {
  int e;
  e= errno;
  fprintf(stderr,"libauthbind's helper: %s: %s\n",m,strerror(e));
  exiterrno(e);
}

static void badusage(void) {
  fprintf(stderr,"libauthbind's helper: bad usage\n");
  exit(ENOSYS);
}

static struct sockaddr_in saddr4;
static struct sockaddr_in6 saddr6;

static struct sockaddr *saddr_any;
static const void *addr_any;
static size_t saddrlen_any, addrlen_any;

static void authorised(void) {
  if (bind(0,saddr_any,saddrlen_any)) exiterrno(errno);
  else _exit(0);
}

static void checkexecflagfile(const char *file) {
  if (!access(file,X_OK)) authorised();
  if (errno != ENOENT) exiterrno(errno);
}

static void hex2bytes(const char *string, unsigned char *out, int len) {
  int i;
  for (i=0; i<len; i++) {
    char hex[3], *ep;
    hex[0]= *string++;  if (!hex[0]) badusage();
    hex[1]= *string++;  if (!hex[1]) badusage();
    hex[2]= 0;
    *out++ = strtoul(hex,&ep,16);
    if (ep != &hex[2]) badusage();
  }
}

int main(int argc, const char *const *argv) {
  uid_t uid;
  char fnbuf[300];
  char *ep;
  const char *np;
  const char *tophalfchar="";
  unsigned long port, addr4=0, haddr4=0;
  unsigned int hport;
  int af;
  FILE *file;

  if (argc == 3) {
    af= AF_INET;
    saddr_any= (void*)&saddr4;
    saddrlen_any= sizeof(saddr4);
    addr_any= &saddr4.sin_addr.s_addr;
    addrlen_any= sizeof(saddr4.sin_addr.s_addr);
    addr4= strtoul(argv[1],&ep,16);
    if (*ep || addr4&~0x0ffffffffUL) badusage();
    haddr4= ntohl(addr4);
  } else if (argc == 4 && !strcmp(argv[3],"6")) {
    af= AF_INET6;
    saddr_any= (void*)&saddr6;
    saddrlen_any= sizeof(saddr6);
    addr_any= &saddr6.sin6_addr.s6_addr;
    addrlen_any= sizeof(saddr6.sin6_addr.s6_addr);
    hex2bytes(argv[1], saddr6.sin6_addr.s6_addr,
	      sizeof(saddr6.sin6_addr.s6_addr));
  } else {
    badusage();
    abort();
  }

  port= strtoul(argv[2],&ep,16); if (*ep || port&~0x0ffffUL) badusage();
  hport= htons(port);
  if (hport >= IPPORT_RESERVED/2) tophalfchar= "!";

  if (chdir(CONFIGDIR)) perrorfail("chdir " CONFIGDIR);

  fnbuf[sizeof(fnbuf)-1]= 0;

  switch (af) {
  case AF_INET:
    saddr4.sin_family= af;
    saddr4.sin_port= port;
    saddr4.sin_addr.s_addr= addr4;
    break;
  case AF_INET6:
    saddr6.sin6_family= af;
    saddr6.sin6_port= port;
    break;
  default:
    abort();
  }

  snprintf(fnbuf,sizeof(fnbuf)-1,"byport/%s%u",tophalfchar,hport);
  if (!access(fnbuf,X_OK)) authorised();
  if (errno != ENOENT) exiterrno(errno);

  char npbuf[INET_ADDRSTRLEN + INET6_ADDRSTRLEN];
  np= inet_ntop(af,addr_any,npbuf,sizeof(npbuf));
  assert(np);

  if (af == AF_INET) {
    snprintf(fnbuf,sizeof(fnbuf)-1,"byaddr/%s%s:%u",tophalfchar,np,hport);
    checkexecflagfile(fnbuf);
  }

  snprintf(fnbuf,sizeof(fnbuf)-1,"byaddr/%s%s,%u",tophalfchar,np,hport);
  checkexecflagfile(fnbuf);

  if (af == AF_INET6) {
    char sbuf[addrlen_any*3+1], *sp = sbuf;
    const unsigned char *ip = addr_any;
    int i;
    for (i=0; i<8; i++) {
      unsigned val = 0;
      val |= *ip++;  val <<= 8;
      val |= *ip++;
      if (i) *sp++ = ':';
      sp += sprintf(sp,"%x",val);
    }
    snprintf(fnbuf,sizeof(fnbuf)-1,"byaddr/%s%s,%u",tophalfchar,sbuf,hport);
    checkexecflagfile(fnbuf);
  }

  uid= getuid(); if (uid==(uid_t)-1) perrorfail("getuid");
  snprintf(fnbuf,sizeof(fnbuf)-1,"byuid/%s%lu",tophalfchar,(unsigned long)uid);

  file= fopen(fnbuf,"r");
  if (!file) exiterrno(errno==ENOENT ? EPERM : errno);

  while (fgets(fnbuf,sizeof(fnbuf)-1,file)) {
    unsigned int a1,a2,a3,a4, alen,pmin,pmax;
    int nchar;

    if (af == AF_INET &&
	(nchar = -1,
	 sscanf(fnbuf," %u.%u.%u.%u/%u: %u,%u %n",
		&a1,&a2,&a3,&a4,&alen,&pmin,&pmax,&nchar),
	 nchar == strlen(fnbuf))) {

      if (alen>32 || pmin&~0x0ffff || pmax&~0x0ffff ||
	  a1&~0x0ff || a2&~0xff || a3&~0x0ff || a4&~0x0ff)
	continue;

      unsigned long thaddr, thmask;
      thaddr= (a1<<24)|(a2<<16)|(a3<<8)|(a4);
      thmask= alen ? 0x0ffffffffUL<<(32-alen) : 0;
      if ((haddr4&thmask) != thaddr) continue;

    } else {

      char *comma = strchr(fnbuf,',');
      if (!comma) continue;
      *comma++ = '\0';

      char *hyphen = strchr(fnbuf,'-');
      const char *min, *max;
      if (hyphen) {
	*hyphen++ = '\0';
	min = fnbuf;
	max = hyphen;
      } else {
	min = fnbuf;
	max = fnbuf;
      }
      unsigned char minaddr[addrlen_any];
      unsigned char maxaddr[addrlen_any];
      if (inet_pton(af,min,minaddr) != 1 ||
	  inet_pton(af,max,maxaddr) != 1)
	continue;
      if (memcmp(addr_any,minaddr,addrlen_any) < 0 ||
	  memcmp(addr_any,maxaddr,addrlen_any) > 0)
	continue;

      nchar = -1;
      sscanf(comma," %u-%u %n",
	     &pmin,&pmax,&nchar);
      if (nchar != strlen(comma))
	continue;

    }
    if (hport<pmin || hport>pmax) continue;

    authorised();
  }
  if (ferror(file)) perrorfail("read per-uid file");
  _exit(ENOENT);
}


authbind.c
/*
 *  authbind.c - main invoker program
 *
 *  authbind is Copyright (C) 1998 Ian Jackson
 * 
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 * 
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software Foundation,
 *  Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. 
 * 
 */

#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <unistd.h>
#include <assert.h>

#include "authbind.h"

static void printusage(FILE *f) {
  if (fprintf(f,
	      "usage:       authbind [<options>] <program> <arg> <arg> ...\n"
	      "options:     --deep    --depth <levels>\n"
	      "version:     " MAJOR_VER "." MINOR_VER "\n")
      == EOF) { perror("printf usage"); exit(-1); }
}

static void usageerror(const char *msg) {
  fprintf(stderr,"usage error: %s\n",msg);
  printusage(stderr);
  exit(-1);
}

static void mustsetenv(const char *var, const char *val) {
  if (setenv(var,val,1)) { perror("authbind: setenv"); exit(-1); }
}

int main(int argc, char *const *argv) {
  const char *expreload, *authbindlib, *preload;
  char *newpreload, *ep;
  char buf[50];
  unsigned long depth;

  depth= 1;
  while (argc>1 && argv[1][0]=='-') {
    argc--; argv++;
    if (!argv[0][1]) break;
    if (!strcmp("--deep",argv[0])) { depth= 0; }
    else if (!strcmp("--depth",argv[0])) {
      if (argc<=1) usageerror("--depth requires a value");
      argc--; argv++;
      depth= strtoul(argv[0],&ep,10);
      if (*ep || depth<=0 || depth>100) usageerror("--depth value is not valid");
    } else if (!strcmp("--help",argv[0]) || !strcmp("--help",argv[0])) {
      printusage(stdout);
      exit(0);
    }
  }
  if (argc<2) usageerror("need program name");

  authbindlib= getenv(AUTHBINDLIB_VAR);
  if (!authbindlib) {
    authbindlib= LIBAUTHBIND;
    mustsetenv(AUTHBINDLIB_VAR,authbindlib);
  }
    
  if ((expreload= getenv(PRELOAD_VAR))) {
    newpreload= malloc(strlen(expreload)+strlen(authbindlib)+2);
    strcpy(newpreload,expreload);
    strcat(newpreload,":");
    strcat(newpreload,authbindlib);
    preload= newpreload;
  } else {
    preload= authbindlib;
  }
  mustsetenv(PRELOAD_VAR,preload);

  if (depth > 1) {
    sprintf(buf,"%ld",depth-1);
    mustsetenv(AUTHBIND_LEVELS_VAR,buf);
  } else if (depth == 0) {
    mustsetenv(AUTHBIND_LEVELS_VAR,"y");
  } else {
    assert(depth==1);
  }

  execvp(argv[1],argv+1);
  perror(argv[1]); exit(-1);
}


authbind.h
/*
 *  authbind.h - various definitions
 *
 *  authbind is Copyright (C) 1998 Ian Jackson
 * 
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 * 
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software Foundation,
 *  Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. 
 * 
 */

#ifndef AUTHBIND_H
#define AUTHBIND_H

#ifndef LIBAUTHBIND
# define LIBAUTHBIND "/usr/local/lib/authbind/libauthbind.so." MAJOR_VER
#endif

#ifndef HELPER
# define HELPER "/usr/local/lib/authbind/helper"
#endif

#ifndef CONFIGDIR
# define CONFIGDIR "/etc/authbind"
#endif

#define PRELOAD_VAR "LD_PRELOAD"
#define AUTHBINDLIB_VAR "AUTHBIND_LIB"
#define AUTHBIND_LEVELS_VAR "AUTHBIND_LEVELS"

static inline void bytes2hex(const unsigned char *in, char *out, size_t len) {
  size_t i;
  for (i=0; i<16; i++)
    sprintf(out+i*2,"%02x", in[i]);
}

#endif


libauthbind.c
/*
 *  libauthbind.c - bind(2)-redirector library for authbind
 *
 *  authbind is Copyright (C) 1998 Ian Jackson
 * 
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 * 
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software Foundation,
 *  Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. 
 * 
 */

#include <dlfcn.h>
#include <stdio.h>
#include <sys/types.h>
#include <unistd.h>
#include <errno.h>
#include <stdlib.h>
#include <signal.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/wait.h>
#include <netinet/in.h>

#include "authbind.h"

typedef void anyfn_type(void);
typedef int bindfn_type(int fd, const struct sockaddr *addr, socklen_t addrlen);

#define STDERRSTR_CONST(m) write(2,m,sizeof(m)-1)
#define STDERRSTR_STRING(m) write(2,m,strlen(m))

static anyfn_type *find_any(const char *name) {
  static const char *dlerr;
  anyfn_type *kv;

  kv= dlsym(RTLD_NEXT,name); if (kv) return kv;
  dlerr= dlerror(); if (!dlerr) dlerr= "dlsym() failed for no reason";
  STDERRSTR_CONST("libauthbind: error finding original version of ");
  STDERRSTR_STRING(name);
  STDERRSTR_CONST(": ");
  STDERRSTR_STRING(dlerr);
  STDERRSTR_STRING("\n");
  errno= ENOSYS;
  return 0;
}

static bindfn_type find_bind, *old_bind= find_bind;

int find_bind(int fd, const struct sockaddr *addr, socklen_t addrlen) {
  anyfn_type *anyfn;
  anyfn= find_any("bind"); if (!anyfn) return -1;
  old_bind= (bindfn_type*)anyfn;
  return old_bind(fd,addr,addrlen);
}

static int exiterrno(int e) {
  _exit(e>0 && e<128 ? e : -1);
}

static void removepreload(void) {
  const char *myself, *found;
  char *newval, *preload;
  int lpreload, lmyself, before, after;

  preload= getenv(PRELOAD_VAR);
  myself= getenv(AUTHBINDLIB_VAR);
  if (!myself || !preload) return;

  lpreload= strlen(preload);
  lmyself= strlen(myself);

  if (lmyself < 1 || lpreload<lmyself) return;
  if (lpreload==lmyself) {
    if (!strcmp(preload,myself)) unsetenv(PRELOAD_VAR);
    return;
  }
  if (!memcmp(preload,myself,lmyself) && preload[lmyself]==':') {
    before= 0; after= lpreload-(lmyself+1);
  } else if (!memcmp(preload+lpreload-lmyself,myself,lmyself) &&
	     preload[lpreload-(lmyself+1)]==':') {
    before= lpreload-(lmyself+1); after= 0;
  } else {
    if (lpreload<lmyself+2) return;
    found= preload+1;
    for (;;) {
      found= strstr(found,myself); if (!found) return;
      if (found > preload+lpreload-(lmyself+1)) return;
      if (found[-1]==':' && found[lmyself]==':') break;
      found++;
    }
    before= found-preload;
    after= lpreload-(before+lmyself+1);
  }
  newval= malloc(before+after+1);
  if (newval) {
    memcpy(newval,preload,before);
    strcpy(newval+before,preload+lpreload-after);
    if (setenv(PRELOAD_VAR,newval,1)) return;
    free(newval);
  }
  strcpy(preload+before,preload+lpreload-after);
  return;
}

int _init(void);
int _init(void) {
  char *levels;
  int levelno;

  /* If AUTHBIND_LEVELS is
   *  unset => always strip from preload
   *  set and starts with `y' => never strip from preload, keep AUTHBIND_LEVELS
   *  set to integer > 1 => do not strip now, subtract one from AUTHBIND_LEVELS
   *  set to integer 1 => do not strip now, unset AUTHBIND_LEVELS
   *  set to empty string or 0 => strip now, unset AUTHBIND_LEVELS
   */
  levels= getenv(AUTHBIND_LEVELS_VAR);
  if (levels) {
    if (levels[0]=='y') return 0;
    levelno= atoi(levels);
    if (levelno > 0) {
      levelno--;
      if (levelno > 0) sprintf(levels,"%d",levelno);
      else unsetenv(AUTHBIND_LEVELS_VAR);
      return 0;
    }
    unsetenv(AUTHBIND_LEVELS_VAR);
  }
  removepreload();
  return 0;
}

static const int evilsignals[]= { SIGFPE, SIGILL, SIGSEGV, SIGBUS, 0 };

int bind(int fd, const struct sockaddr *addr, socklen_t addrlen) {
  pid_t child, rchild;
  char portarg[5], addrarg[33];
  const char *afarg;
  int i, r, status;
  const int *evilsignal;
  sigset_t block, saved;
  unsigned int portval;

  switch (addr->sa_family) {
  case AF_INET:
    portval = ((struct sockaddr_in*)addr)->sin_port;
    if (addrlen != sizeof(struct sockaddr_in)) goto bail;
    break;
  case AF_INET6:
    portval = ((struct sockaddr_in6*)addr)->sin6_port;
    if (addrlen != sizeof(struct sockaddr_in6)) goto bail;
    break;
  default:
    goto bail;
  }

  if (!geteuid() || portval == 0 || ntohs(portval) >= IPPORT_RESERVED) {
  bail:
    return old_bind(fd,addr,addrlen);
  }

  sigfillset(&block);
  for (evilsignal=evilsignals;
       *evilsignal;
       evilsignal++)
    sigdelset(&block,*evilsignal);
  if (sigprocmask(SIG_BLOCK,&block,&saved)) return -1;

  switch (addr->sa_family) {
  case AF_INET:
    afarg = 0;
    sprintf(addrarg,"%08lx",
	    ((unsigned long)(((struct sockaddr_in*)addr)->sin_addr.s_addr))
	    &0x0ffffffffUL);
    break;
  case AF_INET6:
    afarg = "6";
    bytes2hex(((struct sockaddr_in6*)addr)->sin6_addr.s6_addr, addrarg, 16);
    break;
  default:
    abort();
  }
  sprintf(portarg,"%04x",
	  portval&0x0ffff);

  child= fork(); if (child==-1) goto x_err;

  if (!child) {
    if (dup2(fd,0)) exiterrno(errno);
    removepreload();
    execl(HELPER,HELPER,addrarg,portarg,afarg,(char*)0);
    status= errno > 0 && errno < 127 ? errno : 127;
    STDERRSTR_CONST("libauthbind: possible installation problem - "
		    "could not invoke " HELPER "\n");
    exiterrno(status);
  }

  rchild= waitpid(child,&status,0);
  if (rchild==-1) goto x_err;
  if (rchild!=child) { errno= ECHILD; goto x_err; }

  if (WIFEXITED(status)) {
    if (WEXITSTATUS(status)) {
      errno= WEXITSTATUS(status);
      if (errno >= 127) errno= ENXIO;
      goto x_err;
    }
    r= 0;
    goto x;
  } else {
    errno= ENOSYS;
    goto x_err;
  }

x_err:
  r= -1;
x:
  if (sigprocmask(SIG_SETMASK,&saved,0)) abort();
  return r;
}
